CERT-8241: revert in cvl (#354)
Certora/Documentation: docs/cvl/statements.md

In this example, the `@withrevert` modifier is applied to the `Deposit` function call, which is expected to revert if the `value` exceeds the specified `limit`. The `assert` statement checks whether `lastReverted` is true, ensuring that the contract execution does revert as anticipated when the condition is violated. The error message in the `assert` provides additional context about the expectation.

> By default, this applies only for Solidity function calls. With the `--prover_args "-cvlFunctionRevert"` option, CVL functions can revert as well, and the `revert` statement becomes available in CVL. In this mode, CVL functions also set the `lastReverted` variable, just like Solidity functions and reverts of calls without `@withrevert` are propagated up to their callers. Only if no call in the chain had a `@withrevert` annotation do we assume no revert happened. A call with `@withrevert` stops propagation of reverts and will not make the calling function revert, but instead allows reading the `lastReverted` variable immediately after it to check whether the call reverted.

Reviewer: ```suggestion
By default, this applies only for Solidity function calls. With the `--prover_args "-cvlFunctionRevert"` option, CVL functions can revert as well, and the `revert` statement becomes available in CVL. In this mode, CVL functions also set the `lastReverted` variable, just like Solidity functions and reverts of calls without `@withrevert` are propagated up to their callers. Only if no call in the call stack had a `@withrevert` annotation do we assume no revert happened. A call with `@withrevert` stops propagation of reverts and will not make the calling function revert, but instead allows reading the `lastReverted` variable immediately after it to check whether the call reverted.
```

> By default, this applies only for Solidity function calls. With the `--prover_args "-cvlFunctionRevert"` option, CVL functions can revert as well, and the `revert` statement becomes available in CVL. In this mode, CVL functions also set the `lastReverted` variable, just like Solidity functions and reverts of calls without `@withrevert` are propagated up to their callers. Only if no call in the call stack had a `@withrevert` annotation do we assume no revert happened. A call with `@withrevert` stops propagation of reverts and will not make the calling function revert, but instead allows reading the `lastReverted` variable immediately after it to check whether the call reverted.

Reviewer 2: If this option is user facing, it should have its own flag and not just a Prover arg. Then it would have better validation, better error messages, and appear in `--help`.
creating a flag from a prover arg is very easy, you can see an example by looking for the string `OPTIMISTIC_HASHING` in the file `scripts/CertoraProver/certoraContextAttributes.py`

A CVL function can revert either from a call inside it (without `@withrevert`) that reverts or from an explicit revert using the `revert` statement.

##### Example:

```cvl
function cvlFunctionThatMayRevert(bool input) {
    if (!input) {
        revert("Input was false in CVL function");
    }
}

rule revertInCVL {
    bool b;
    cvlFunctionThatMayRevert@withrevert(b);
    assert lastReverted <=> !b;
}
```

> - [Further examples](https://github.com/Certora/Examples/blob/ae2eca20d8e6caf378ff10cf8066ecfc45d3658d/CVLByExample/RevertKeyWord/example.spec)

Reviewer: Let's rather link to the example location in the repository assuming the PR was merged already. As you say, the link may not be available after merging.

Author: I did it like this because all the links in the page where such blob links, so I wasn't sure what's the usual process - do we just merge the Examples PR first and then update the link after?

In this example, the `@withrevert` modifier is applied to the `Deposit` function call, which is expected to revert if the `value` exceeds the specified `limit`. The `assert` statement checks whether `lastReverted` is true, ensuring that the contract execution does revert as anticipated when the condition is violated. The error message in the `assert` provides additional context about the expectation.

> Since version TODO, this applies not only for Solidity function calls. CVL functions can now revert as well, and the `revert` statement is available in CVL. CVL functions also set the `lastReverted` variable, just like Solidity functions and reverts of calls without `@withrevert` are propagated up to their callers. Only if no call in the call stack had a `@withrevert` annotation do we assume no revert happened. A call with `@withrevert` stops propagation of reverts and will not make the calling function revert, but instead allows reading the `lastReverted` variable immediately after it to check whether the call reverted.

Reviewer: ```suggestion
Since version 8, this applies not only for Solidity function calls. CVL functions can now revert as well, and the `revert` statement is available in CVL. CVL functions also set the `lastReverted` variable, just like Solidity functions and reverts of calls without `@withrevert` are propagated up to their callers. Only if no call in the call stack had a `@withrevert` annotation do we assume no revert happened. A call with `@withrevert` stops propagation of reverts and will not make the calling function revert, but instead allows reading the `lastReverted` variable immediately after it to check whether the call reverted.
```

```
- [Further examples](https://github.com/Certora/Examples/blob/ae2eca20d8e6caf378ff10cf8066ecfc45d3658d/CVLByExample/RevertKeyWord/example.spec)

```{note}

> It is possible to enable the old behavior where CVL functions cannot revert and do not set the `lastReverted` variable with the `--prover_args "-cvlFunctionRevert false"` option. This is meant as a compatibility option where specs need to be adjusted to the new behavior and will be retired in a future release.

Reviewer: ```suggestion
It is possible to enable the behavior before version TODO where CVL functions cannot revert and do not set the `lastReverted` variable with the `--prover_args "-cvlFunctionRevert false"` option. This is meant as a compatibility option where specs need to be adjusted to the new behavior and will be retired in a future release.
```
